### ISSUE TITLE
Thumbnails for people do not show under People (Face Recognition)

### ISSUE THREAD

**Issue by MikeDuncan1** [pulsejet/memories#495]
**Describe the bug**
Thumbnails do not show up for People(face recognition)
**To Reproduce**
Properly install facerecognition and let it scan

**Screenshots**
https://imgur.com/a/aKxDdgW

**Platform (please complete the following information):**

- OS: Android and Windows
- Browser Chrome, Edge and Firefox
- Version Latest

**Additional context**
It kind of sucks XD

**Important: Are all of the following are true?**

- I have followed the [configuration steps](https://github.com/pulsejet/memories/wiki/Configuration).
- I am NOT using `OCCWeb` to run `occ` commands
- The problem persists after clearing browser cache
true true true
![chrome_qEn726u6Zv](https://user-images.githubusercontent.com/29655039/226057567-fc455624-0c7b-42b2-bfc6-8d15d5e2573e.png)

**Comment by pulsejet** [comment]
Anything in logs?

**Comment by MikeDuncan1** [comment]
I wasn't sure where to start looking honestly.  The thumbnails show up in the backend for face recognition and every other thumbnail works.  What log should I pull that might help?  Thanks for looking

**Comment by pulsejet** [comment]
Nextcloud logs, as well as JS console / network tab

**Comment by MikeDuncan1** [comment]
![chrome_aVgXgBG9RQ](https://user-images.githubusercontent.com/29655039/226075628-9a3986d3-a468-4d43-9725-fd083c2a2e5e.png)

Will get back on this in a bit, here is what it is putting in place of the picture, I assume that's a base64 encoded placeholder gif?

**Comment by MikeDuncan1** [comment]
![A2GkpqO](https://user-images.githubusercontent.com/29655039/226088247-dee75047-97a8-440d-8c73-b45b0b1496a2.png)
![HReAeqA](https://user-images.githubusercontent.com/29655039/226088248-d27769fe-cfbe-4a82-8fbb-230d0203714d.png)

**Comment by MikeDuncan1** [comment]
[nextlog.txt](https://github.com/pulsejet/memories/files/11007651/nextlog.txt)

**Comment by pulsejet** [comment]
Any idea @matiasdelellis?

**Comment by matiasdelellis** [comment]
Hi @MikeDuncan1 
Let's start from some base.  What version of nextcloud do you use? Database? Face recognition?.

The photos are all local.?  In user preferences? Do you see the faces?

**Comment by MikeDuncan1** [comment]
The only thing I can think of since posting is that I did enable "pretty permalinks" so it would not show index.php all the time. I followed a guide for nextcloud specifically and everything seems to work correctly. I don't know if this could cause an issue though

**Comment by matiasdelellis** [comment]
Hey... 😅 

> Everything seems to work correctly. 

This too?
There are possibilities that it is related, I don't remember how the links are generated here, but I had problems before because of it.

**Comment by MikeDuncan1** [comment]
> Hi @MikeDuncan1 Let's start from some base. What version of nextcloud do you use? Database? Face recognition?.
> 
> The photos are all local.? In user preferences? Do you see the faces?

Ok nextcloud version 25.0.4

Face recognition 0.9.11

Mariadb

I can see the thumbnails just fine in settings/user/facerecognition

**Comment by MikeDuncan1** [comment]
![hyOZnWU](https://user-images.githubusercontent.com/29655039/226145619-c38380f7-164c-42a9-939a-4aa2e4524eef.png)
from /settings/user/facerecognition, the thumbnail url is different there

**Comment by MikeDuncan1** [comment]
> Hey... 😅
> 
> > Everything seems to work correctly.
> 
> This too? There are possibilities that it is related, I don't remember how the links are generated here, but I had problems before because of it.

Thumbnails in memories for face recognition is the only thing not working, let me find the instructions I used

**Comment by MikeDuncan1** [comment]
https://docs.nextcloud.com/server/21/admin_manual/installation/source_installation.html?highlight=htaccess%20rewritebase#pretty-urls

This is what I followed, my install is not in a subfolder

**Comment by matiasdelellis** [comment]
Nop.. . Seems like it's not related. In my case, I disabled the pretty urls and it still works for all my accounts. 🤔

Reviewing the code, a very unlikely option would be that these photos are in a folder that does not handle Memories, but you would not see the images in the timeline view of Memories. So, When you click on the people, do you see the images?

I forgot to ask about the memories version. I'm trying to reproduce it with the same version, but it's strange. 😞

**Comment by MikeDuncan1** [comment]
Memories version is the latest one that was released yesterday, I am able to see all the photos in the timeline and also I can see the photos when I tap on the person under people face recognition. The only thing that is missing is the thumbnails

**Comment by MikeDuncan1** [comment]
Thanks for the help by the way and if you want an admin login or for me to do anything let me know. The install is on its own vm and I can back it up really fast and change anything at all with no repercussions

**Comment by matiasdelellis** [comment]
Hi @MikeDuncan1 
Well, You can send me a test account with administrator permissions to mati86 at gmail com.. There is no need to make any changes, the permissions would be only to be able to access the nextcloud logs.  🤦🏻 

I think it will be faster to avoid continuing to ask you useless questions and try directly. 😅

**Comment by MikeDuncan1** [comment]
It is done, thank you :)  Let me know if I need to do any more or run anything on the server

**Comment by matiasdelellis** [comment]
![imagen](https://user-images.githubusercontent.com/733715/226484739-6b971d1e-fde2-4c9d-a352-647f446964b4.png)

mati86dl at gmail com

Sorry.. 😞

**Comment by MikeDuncan1** [comment]
Haha ok sent to that email

**Comment by matiasdelellis** [comment]
Well, 
![imagen](https://user-images.githubusercontent.com/733715/226492702-b1c2508d-60fb-4311-937b-e333a1a4755b.png)

Thanks to @MikeDuncan1 I was able to reproduce it very quickly, but I still have to investigate better. 🤔

**Comment by MikeDuncan1** [comment]
Thank you, whatever your need next let me know, hopefully I don't have to rebuild this thing from the ground up lmao

**Comment by matiasdelellis** [comment]
Ok, For now, I can comment that it is a problem in the backend/controller and not the Memories fronted.

Downloading the picture of a cluster that i see in Memories.
```
[matias@bangho-mdl Escritorio]$ curl 'https://cloud.***.com/apps/memories/api/facerecognition/people/preview/257' -u matiasdelellis:passsword -o preview.png
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  400k  100  400k    0     0   144k      0  0:00:02  0:00:02 --:--:--  144k
[matias@bangho-mdl Escritorio]$ xdg-open preview.png 
```
![imagen](https://user-images.githubusercontent.com/733715/226494448-f2034195-873a-47d6-b057-8cededbeec58.png)


And trying to download a missing image.. 
```
[matias@bangho-mdl Escritorio]$ curl 'https://cloud.**.com/apps/memories/api/facerecognition/people/preview/251' -u matiasdelellis:password -o preview.png
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100     2  100     2    0     0      1      0  0:00:02  0:00:01  0:00:01     1
[matias@bangho-mdl Escritorio]$ cat preview.png 
[]
```

Just download [] with 404.. 🤦🏻‍♂️

So, we can discard any type of cache or timeout problem. 🤔

I'm going to review that code to see if I can think something, but in the [next version of memories](https://github.com/pulsejet/memories/commit/5f9a299af4224363b3df8c69c3e6d4433681c3b7) there will be a much better log who can help us here.

**Comment by pulsejet** [comment]
My best guess is that `getFaceRecognitionPreview` is returning empty for some reason

**Comment by matiasdelellis** [comment]
> My best guess is that `getFaceRecognitionPreview` is returning empty for some reason

Agree. 🤔 

@MikeDuncan1 

You can run this query?
It is a simplified version of the related query. (It doesn't use the magic of Memories with the cache.)

```
SELECT frf.x,
frf.y,
frf.width,
frf.height,
fri.file as file_id,
m.w as image_width,
m.h as image_height,
frf.confidence,
m.fileid,
m.datetaken
FROM `oc_facerecog_faces` AS `frf`
INNER JOIN `oc_facerecog_images`  `fri` ON fri.id=frf.image AND fri.model=4
INNER JOIN `oc_facerecog_persons`  `frp` ON frp.id=frf.person
INNER JOIN  `oc_memories`  `m` ON m.fileid=fri.file
WHERE fri.user='matiasdelellis'
AND frp.name='Rajesh'
LIMIT 15
```

And
```
SELECT frf.x,
frf.y,
frf.width,
frf.height,
fri.file as file_id,
m.w as image_width,
m.h as image_height,
frf.confidence,
m.fileid,
m.datetaken
FROM `oc_facerecog_faces` AS `frf`
INNER JOIN `oc_facerecog_images`  `fri` ON fri.id=frf.image AND fri.model=4
INNER JOIN `oc_facerecog_persons`  `frp` ON frp.id=frf.person
INNER JOIN  `oc_memories`  `m` ON m.fileid=fri.file
WHERE fri.user='matiasdelellis'
AND frp.id=249
LIMIT 15
```

**Comment by MikeDuncan1** [comment]
do I just open mariadb from cli and drop this in?  will do it when I get home in a minute

**Comment by matiasdelellis** [comment]
Yes.. You have to select the database first, but essentially that.. 😉 

`use nextcloud`

If your database is different that `nextcloud` use the correct name.

**Comment by matiasdelellis** [comment]
It doesn't solve anything... it's just to evaluate the query.

```
MariaDB [nextcloud]> SELECT frf.x, frf.y, frf.width, frf.height, fri.file as file_id, m.w as image_width, m.h as image_height, frf.confidence, m.fileid, m.datetaken FROM `oc_facerecog_faces` AS `frf` INNER JOIN `oc_facerecog_images`  `fri` ON fri.id=frf.image AND fri.model=4 INNER JOIN `oc_facerecog_persons`  `frp` ON frp.id=frf.person INNER JOIN  `oc_memories`  `m` ON m.fileid=fri.file WHERE fri.user='user' AND frp.name='Rajesh' LIMIT 15;
+------+-----+-------+--------+---------+-------------+--------------+-----------------+--------+---------------------+
| x    | y   | width | height | file_id | image_width | image_height | confidence      | fileid | datetaken           |
+------+-----+-------+--------+---------+-------------+--------------+-----------------+--------+---------------------+
| 1097 | 203 |   115 |    116 |  101616 |        1500 |         1000 | 1.0460413694382 | 101616 | 2019-11-21 15:05:54 |
|  241 | 478 |   115 |    115 |  100826 |        1500 |          999 | 1.0245764255524 | 100826 | 2019-11-21 03:52:27 |
|  312 | 113 |   110 |    111 |  101567 |        1200 |          800 | 1.0742532014847 | 101567 | 2019-11-21 15:05:46 |
|  683 | 203 |   105 |    106 |  100668 |        1244 |          700 | 1.0702662467957 | 100668 | 2019-11-21 03:52:14 |
|  805 | 387 |    91 |     90 |  100672 |        1400 |          950 |  1.048597574234 | 100672 | 2019-11-21 03:52:14 |
| 1359 | 272 |   199 |    198 |  100846 |        3000 |         2122 | 1.0510174036026 | 100846 | 2019-11-21 03:52:28 |
|  981 | 155 |    63 |     64 |  100699 |        1200 |          800 |  1.098183631897 | 100699 | 2019-11-21 03:52:17 |
|  678 | 286 |   128 |    128 |  101533 |        2048 |         1280 | 1.0477367639542 | 101533 | 2019-11-21 15:05:37 |
|  884 | 147 |    85 |     85 |  100704 |        1200 |          675 | 1.0349026918411 | 100704 | 2019-11-21 03:52:17 |
|  904 | 215 |    61 |     60 |  100700 |        1400 |          875 | 1.0258586406708 | 100700 | 2019-11-21 03:52:17 |
|  525 | 277 |   123 |    122 |  100662 |        1200 |          675 | 1.0733910799026 | 100662 | 2019-11-21 03:52:13 |
| 1735 | 188 |   170 |    170 |  100750 |        2000 |         1125 | 1.0526965856552 | 100750 | 2019-11-21 03:52:20 |
|  333 | 111 |   102 |    102 |  101619 |        1000 |          567 | 1.0465307235718 | 101619 | 2019-11-21 15:05:54 |
|  642 | 352 |   118 |    119 |  100778 |        2000 |         1130 | 1.0499882698059 | 100778 | 2019-11-21 03:52:23 |
|  707 | 285 |    74 |     74 |  100657 |        1500 |          844 |  1.060086607933 | 100657 | 2019-05-01 17:19:23 |
+------+-----+-------+--------+---------+-------------+--------------+-----------------+--------+---------------------+
15 rows in set (0,018 sec)
```

You should see a table like this. 🤔

**Comment by MikeDuncan1** [comment]
Was each of those supposed to be one line/command? Will go back in a few, sorry not a lot of experience running queries directly besides initial db creation

**Comment by MikeDuncan1** [comment]
This is what I got back out of it:

MariaDB [nextcloud]> SELECT frf.x, frf.y, frf.width, frf.height, fri.file as file_id, m.w as image_width, m.h as image_height, frf.confidence, m.fileid, m.datetaken FROM 'oc_facerecog_faces' AS 'frf' INNER JOIN 'oc_facerecog_images'  'fri' ON fri.id=frf.image AND fri.model=4 INNER JOIN 'oc_facerecog_persons'  'frp' ON frp.id=frf.person INNER JOIN  'oc_memories'  'm' ON m.fileid=fri.file WHERE fri.user='user' AND frp.name='Rajesh' LIMIT 15;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ''oc_facerecog_faces' AS 'frf' INNER JOIN 'oc_facerecog_images'  'fri' ON fri....' at line 1

**Comment by MikeDuncan1** [comment]
Ok I can see that username is not right for here anyway, let me try again with what you initially posted, gotta correct the ' and get it all in one line real quick, when i copied, it copied it as multiple lines

**Comment by MikeDuncan1** [comment]
Ok still not having any success, maybe I'm just formatting the query wrong?

MariaDB [nextcloud]> SELECT frf.x,frf.y,frf.width,frf.height,fri.file as file_id,m.w as image_width,m.h as image_height,frf.confidence,m.fileid,m.datetaken FROM 'oc_facerecog_faces' AS 'frf' INNER JOIN 'oc_facerecog_images' 'fri' ON fri.id=frf.image AND fri.model=4 INNER JOIN 'oc_facerecog_persons' 'frp' ON frp.id=frf.person INNER JOIN 'oc_memories' 'm' ON m.fileid=fri.file WHERE fri.user='matiasdelellis' AND frp.name='Rajesh' LIMIT 15;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ''oc_facerecog_faces' AS 'frf' INNER JOIN 'oc_facerecog_images' 'fri' ON fri.i...' at line 1

**Comment by pulsejet** [comment]
You need backticks for table names, not inverted commas. ` instead of '

**Comment by MikeDuncan1** [comment]
Ok, I pulled it together tonight and here is the output for the first query:

MariaDB [nextcloud]> SELECT frf.x,frf.y,frf.width,frf.height,fri.file as file_id,m.w as image_width,m.h as image_height,frf.confidence,m.fileid,m.datetaken FROM `oc_facerecog_faces` AS `frf` INNER JOIN `oc_facerecog_images`  `fri` ON fri.id=frf.image AND fri.model=4 INNER JOIN `oc_facerecog_persons`  `frp` ON frp.id=frf.person INNER JOIN  `oc_memories`  `m` ON m.fileid=fri.file WHERE fri.user='matiasdelellis' AND frp.name='Rajesh' LIMIT 15;
+-----+-----+-------+--------+---------+-------------+--------------+-----------------+--------+---------------------+
| x   | y   | width | height | file_id | image_width | image_height | confidence      | fileid | datetaken           |
+-----+-----+-------+--------+---------+-------------+--------------+-----------------+--------+---------------------+
| 428 | 344 |   257 |    257 |   14998 |        2000 |         1125 | 1.0926768779755 |  14998 | 2019-11-21 03:52:27 |
| 182 | 251 |   171 |    171 |   14900 |        1600 |          900 | 1.0645594596863 |  14900 | 2019-11-21 03:52:26 |
| 741 | 132 |   137 |    136 |   18132 |        1280 |          720 | 1.0608425140381 |  18132 | 2020-07-09 02:14:12 |
+-----+-----+-------+--------+---------+-------------+--------------+-----------------+--------+---------------------+
3 rows in set (0.001 sec)

second query:

MariaDB [nextcloud]> SELECT frf.x,frf.y,frf.width,frf.height,fri.file as file_id,m.w as image_width,m.h as image_height,frf.confidence,m.fileid,m.datetaken FROM `oc_facerecog_faces` AS `frf` INNER JOIN `oc_facerecog_images`  `fri` ON fri.id=frf.image AND fri.model=4 INNER JOIN `oc_facerecog_persons`  `frp` ON frp.id=frf.person INNER JOIN  `oc_memories`  `m` ON m.fileid=fri.file WHERE fri.user='matiasdelellis' AND frp.id=249 LIMIT 15;
+------+------+-------+--------+---------+-------------+--------------+------------------+--------+---------------------+
| x    | y    | width | height | file_id | image_width | image_height | confidence       | fileid | datetaken           |
+------+------+-------+--------+---------+-------------+--------------+------------------+--------+---------------------+
|  383 |  331 |   128 |    129 |   13734 |        1200 |          675 | 0.93576633930206 |  13734 | 2019-11-21 03:52:13 |
|  847 |  475 |   164 |    166 |   14079 |        1200 |          781 |  1.0610724687576 |  14079 | 2019-11-21 03:52:17 |
| 1234 |  405 |   247 |    247 |   14871 |        1920 |         1080 |  1.0697635412216 |  14871 | 2019-11-21 03:52:26 |
|  504 | 1473 |   399 |    398 |   14564 |        2393 |         3266 | 0.90252828598022 |  14564 | 2019-11-21 03:52:22 |
| 1087 |  413 |   189 |    190 |   15403 |        1777 |          999 |  1.0636808872223 |  15403 | 2019-11-21 15:05:42 |
+------+------+-------+--------+---------+-------------+--------------+------------------+--------+---------------------+
5 rows in set (0.005 sec)

**Comment by ngocleek** [comment]
Try this `occ memories:index -f` to see if it runs properly. I was able to solve mine when I fix exiftool path not correct in Nextcloud config.

`'memories.exiftool' => 'exiftool',`

https://github.com/pulsejet/memories/wiki/System-config-options

**Comment by MikeDuncan1** [comment]
> Try this `occ memories:index -f` to see if it runs properly. I was able to solve mine when I fix exiftool path not correct in Nextcloud config.
> 
> `'memories.exiftool' => 'exiftool',`
> 
> https://github.com/pulsejet/memories/wiki/System-config-options

Thanks, I will take a look when I get home

**Comment by pulsejet** [comment]
Try the latest version, that'll give some better errors.

**Comment by MikeDuncan1** [comment]
> Try the latest version, that'll give some better errors.

Just installed :)

**Comment by matiasdelellis** [comment]
Ok.. Keep failing... But with better info.. 😉

> message | "Not found (preview from photos list)"

But with the latest refactoring it's hard to follow the code for me... 😞

Based on the tables that prints the main query for "Face recognition", it's consistent with the images that the cluster has, but when join with the "Memories" table, it ends up returning an empty table.

Do you see anything relevant in the table?
How can we get Michael to do the complete query?

I just realized that we consult for oc_facerecog_images.file and it is unnecessary, but it would not fail for that.

**Comment by MikeDuncan1** [comment]
> Ok.. Keep failing... But with better info.. 😉
> 
> > message | "Not found (preview from photos list)"
> 
> But with the latest refactoring it's hard to follow the code for me... 😞
> 
> Based on the tables that prints the main query for "Face recognition", it's consistent with the images that the cluster has, but when join with the "Memories" table, it ends up returning an empty table.
> 
> Do you see anything relevant in the table? How can we get Michael to do the complete query?
> 
> I just realized that we consult for oc_facerecog_images.file and it is unnecessary, but it would not fail for that.

I can run whatever you need me to run, just let me know

**Comment by MikeDuncan1** [comment]
> Try the latest version, that'll give some better errors.

![Screenshot_20230324-232942.png](https://user-images.githubusercontent.com/29655039/227696095-a2f04d93-d1dc-4d9a-97b3-49e0d1409453.png)

This is new, broken image icons now, never were before

**Comment by MikeDuncan1** [comment]
@matiasdelellis @pulsejet 

[PHP] Error: Error: Undefined array key "name" at /var/www/cloud.myduncans.com/apps/memories/lib/ClustersBackend/FaceRecognitionBackend.php#130 at <<closure>>

0. /var/www/cloud.myduncans.com/apps/memories/lib/ClustersBackend/FaceRecognitionBackend.php line 130
   OC\Log\ErrorHandler::onAll()
1. /var/www/cloud.myduncans.com/apps/memories/lib/Controller/ClustersController.php line 48
   OCA\Memories\ClustersBackend\FaceRecognitionBackend->getClusters()
2. /var/www/cloud.myduncans.com/apps/memories/lib/UtilController.php line 20
   OCA\Memories\Controller\ClustersController->OCA\Memories\Controller\{closure}("*** sensitive parameters replaced ***")
3. /var/www/cloud.myduncans.com/apps/memories/lib/Controller/ClustersController.php line 51
   OCA\Memories\Util::guardEx()
4. /var/www/cloud.myduncans.com/lib/private/AppFramework/Http/Dispatcher.php line 225
   OCA\Memories\Controller\ClustersController->list()
5. /var/www/cloud.myduncans.com/lib/private/AppFramework/Http/Dispatcher.php line 133
   OC\AppFramework\Http\Dispatcher->executeController()
6. /var/www/cloud.myduncans.com/lib/private/AppFramework/App.php line 172
   OC\AppFramework\Http\Dispatcher->dispatch()
7. /var/www/cloud.myduncans.com/lib/private/Route/Router.php line 298
   OC\AppFramework\App::main()
8. /var/www/cloud.myduncans.com/lib/base.php line 1047
   OC\Route\Router->match()
9. /var/www/cloud.myduncans.com/index.php line 36
   OC::handleRequest()

GET /apps/memories/api/clusters/facerecognition
from 192.168.0.245 by matiasdelellis at 2023-03-24T10:27:54+00:00

Seeing this error a lot in the log under /settings/admin/logging

**Comment by sirdrakeistda** [comment]
@MikeDuncan1 do you have also recognize installed? I had the same problem, but not only for facerecognition. It affected recognize too. For me it seems like some image formats cause this problem. png images work well, while jpg cause some issues.
https://github.com/pulsejet/memories/issues/230

**Comment by pulsejet** [comment]
I just realized something -- do you have the PHP Imagemagick extension installed and enabled? @MikeDuncan1

**Comment by MikeDuncan1** [comment]
> I just realized something -- do you have the PHP Imagemagick extension installed and enabled? @MikeDuncan1 

Ok it was installed but extension=imagick.so was not in Apache or cli php.ini file

I've added and restarted now, is there a way to retrigger generation or are you on the right track

**Comment by MikeDuncan1** [comment]
> I just realized something -- do you have the PHP Imagemagick extension installed and enabled? @MikeDuncan1 

![MVIMG_20230326_193638.jpg](https://user-images.githubusercontent.com/29655039/227815344-abd7588f-09f7-402e-90d1-fdee7ca28e44.jpg)

Not broken images anymore :)

**Comment by MikeDuncan1** [comment]
@matiasdelellis  should I rescan all my photos to see if it populates now?

**Comment by Wastl2IT** [comment]
i had the same situation like the last shown picture after updating to nc26/memories 4.12.5 .
after opening the photos-app, waiting for loading the thumbs and then opening memories again, the thumbs were back ;)

**Comment by sirdrakeistda** [comment]
I have no package called imaginary. I have php8.1-imagick installed. I use the nextcloudpi image and I haven't found any hint in the config.php that an imaginary package is used. It worked for me in the past and stopped working from one day to another. I haven't changed anything.

**Comment by MikeDuncan1** [comment]
is there a folder where we can go and look that should contain the previews?

**Comment by MikeDuncan1** [comment]
After updating to 4.13.1 it all works now, thank you for all your efforts :)
